Validation function for ScoreDefinition

shrutiparabgoogle opened the pull request:
Validation logic details are here:https://github.com/apigee/registry/issues/515#issuecomment-1101597784

Note: There are multiple commits in this PR due to merge conflicts while rebasing.  Most of the core changes are in the fourth commit [Add a validation function for ScoreDefinition](https://github.com/apigee/registry/pull/536/commits/3aed6c2221e4d767389a9d6d2e43c7679662e951)

seaneganx commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,153 @@
+		return thresholds[i].GetRange().GetMin() < thresholds[j].GetRange().GetMin()
+	})
+
+	// This loop runs a pointer through the sorted ranges and checks for overlaps, missed values or out of bound scenarios.
```
I'm having some trouble following the algorithm. Is it possible to simplify?

Something like this might work.
1. Check if any of the thresholds have nil ranges. Do we need this step?
2. Sort by range.min
3. Assert the lowest threshold minimum == minValue
4. For each pair of thresholds in the list, assert that left.Max == right.Min - 1
```
for i := 1; i < len(thresholds); i++ {
	left, right := thresholds[i - 1].GetRange(), thresholds[i].GetRange()
	if left.GetMax() < right.GetMin() - 1 {
		// Handle gap.
	} else if left.GetMax() > right.GetMin() - 1 {
		// Handle overlap.
	}
}
```
5. Assert the highest threshold maximum == maxValue

shrutiparabgoogle replied:
Done.

seaneganx reviewed (changes requested):
Looks good for the most part, but I suspect there's a bug so I'm blocking until that's verified as working.

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,153 @@
+		thresholdErrs = append(thresholdErrs, errs...)
+	}
+
+	errs := make([]error, len(patternErrs)+len(formulaErrs)+len(thresholdErrs))
```
Thanks for catching this! 

Yes, it is a bug, I meant to add `len(patternErrs)+len(formulaErrs)+len(thresholdErrs)` as the capacity as you are suggesting in the go playground. I will fix and add modify the tests as suggested in the review.

shrutiparabgoogle replied:
Done

shrutiparabgoogle commented on cmd/registry/scoring/parser_test.go:
```diff
@@ -0,0 +1,911 @@
+	}
+}
+
+func TestValidateScoreDefinitionError(t *testing.T) {
```
Done.

shrutiparabgoogle commented on cmd/registry/scoring/parser_test.go:
```diff
@@ -0,0 +1,911 @@
+}
+
+func TestValidateNumberThresholdsError(t *testing.T) {
+	tests := []struct {
```
Added a couple more conditions:
- Check that all the oneOf fields are set (`formula` and `type`).
- Check that `min_value` < `max_value` in `IntegerType` definition
- Check that `min` <= `max` in `NumberThreshold.Range`

For the two cases that you mentioned:
- `a threshold has minValue > maxValue` -> the added check should take care of this.
-  `the artifact has incomplete information like unspecified/nil number ranges in a threshold` -> this should be ideally taken care of in the proto. We can make the range field required (currently it's not):
```
message NumberThreshold {
  // Represents the severity level of the value of the score.
  Severity severity = 1 [(google.api.field_behavior) = REQUIRED];

  message NumberRange {
    int32 min = 1 [(google.api.field_behavior) = REQUIRED];
    int32 max = 2 [(google.api.field_behavior) = REQUIRED];
  }

  // The boundaries specified in this field will be included in the mentioned
  // severity level. For examples see: PercentType, IntegerType.
  NumberRange range = 2; [(google.api.field_behavior) = REQUIRED];
}
```

This will be inline with how the BooleanThreshold is designed:
```
message BooleanThreshold {
  // Represents the severity level of the value of the score.
  Severity severity = 1 [(google.api.field_behavior) = REQUIRED];

  // The value field will be used to determine the severity
  // level of the generated score.
  bool value = 2 [(google.api.field_behavior) = REQUIRED];
}
```

seaneganx replied:
Do field annotations enforce anything? I was under the impression that we need to validate manually to make sure our annotations are being followed, i.e. they're documentation. I'm not 100% sure though.

shrutiparabgoogle replied:
You are right, the field annotations don't enforce anything. (I just tried to upload a ScoreDefinition which violated a required field).

I thought that they served the same purpose as writing `required string name` in proto2, but it's not true. We will have to manually check the existence of required fields.

shrutiparabgoogle replied:
`the artifact has incomplete information like unspecified/nil number ranges in a threshold` -> in this case, `NumberThreshold.GetRange().GetMin()` will return default value `0` and it will be similar to the case where `min` and `max` in a range are not set. 

I added a test case for this particular case: "missing range"

seaneganx reviewed (commented):
I still have concerns about the correctness, and to a lesser degree, the simplicity of validateNumberThresholds. The requirements are complicated if we want useful and accurate errors for our users, so I tried to incorporate my suggestions into actual code.

```
if len(thresholds) == 0 {
	return nil // Should this be an error?
}

if minValue >= maxValue {
	return []error{
		fmt.Errorf("invalid min_value (%d) and max_value (%d): min_value must be less than max_value", minValue, maxValue)
	}
}

gotMin := maxValue + 1
gotMax := minValue - 1
for _, t := range thresholds {
	r := t.GetRange()
	if r.GetMin() > r.GetMax() {
		errs = append(errs, fmt.Errorf("invalid threshold [%d, %d]: min must not be greater than max", r.GetMin(), r.GetMax()))
		continue // It doesn't really make sense to boundary check an invalid threshold range.
	}

	if r.GetMin() < minValue {
		errs = append(errs, fmt.Errorf("invalid threshold [%d, %d]: min must not be less than min_value (%d)", r.GetMin(), r.GetMax(), minValue))
	} else if r.GetMin() > maxValue {
		errs = append(errs, fmt.Errorf("invalid threshold [%d, %d]: min must not be greater than max_value (%d)", r.GetMin(), r.GetMax(), maxValue))
	}

	if r.GetMax() > maxValue {
		errs = append(errs, fmt.Errorf("invalid threshold [%d, %d]: max must not be greater than max_value (%d)", r.GetMin(), r.GetMax(), maxValue))
	} else if r.GetMax() < minValue {
		errs = append(errs, fmt.Errorf("invalid threshold [%d, %d]: max must not be less than min_value (%d)", r.GetMin(), r.GetMax(), minValue))
	}

	if r.GetMin() < gotMin {
		gotMin = r.GetMin()
	}
	if r.GetMax() > gotMax {
		gotMax = r.GetMax()
	}
}

// Don't check for missing coverage or threshold overlaps unless all the ranges are valid and in bounds.
if len(errs) > 0 {
	return errs
}

if gotMin > minValue {
	errs = append(errs, fmt.Errorf("incomplete coverage: missing coverage for range [%d, %d]", minValue, gotMin - 1))
}
if gotMax < maxValue {
	errs = append(errs, fmt.Errorf("incomplete coverage: missing coverage for range [%d, %d]", gotMax + 1, maxValue))
}

sort.Slice(thresholds, func(i, j int) bool {
	return thresholds[i].GetRange().GetMin() < thresholds[j].GetRange().GetMin()
})

for i := 1; i < len(thresholds); i++ {
	left, right := thresholds[i - 1].GetRange(), thresholds[i].GetRange()
	if left.GetMax() < right.GetMin() - 1 {
		errs = append(errs, fmt.Errorf("incomplete coverage: missing coverage between %d and %d", left.GetMax(), right.GetMin()))
	} else if left.GetMax() > right.GetMin() - 1 {
		errs = append(errs, fmt.Errorf("invalid thresholds [%d, %d] and [%d, %d]: thresholds must not overlap", left.GetMin(), left.GetMax(), right.GetMin(), right.GetMax()))
	}
}
```

This is what I came up with. I didn't actually run this code, so I don't know if it works, but I think it addresses my most recent concerns about returning correct and informative errors. Do you see anything I may have missed?

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,160 @@
+				formulaErrs = append(formulaErrs, errs...)
+			}
+		default:
+			formulaErrs = append(formulaErrs, fmt.Errorf("missing formula, either 'score_formula' or 'rollup_formula' should be set"))
```
There is a testcase for this "missing formula" in `TestValidateScoreDefinition`

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,160 @@
+	})
+
+	// Check coverage for minValue
+	if len(thresholds) > 0 && thresholds[0].GetRange().GetMin() != minValue {
```
Done.

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,160 @@
+	// Check coverage for maxValue
+	if len(thresholds) > 0 && thresholds[len(thresholds)-1].GetRange().GetMax() != maxValue {
+		errs = append(errs, fmt.Errorf("missing coverage for max_value(%d) in threshold value: %q", maxValue, thresholds[len(thresholds)-1]))
+	}
```
Good catch, refactored as suggested above.

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,160 @@
+
+	// Check coverage for minValue
+	if len(thresholds) > 0 && thresholds[0].GetRange().GetMin() != minValue {
+		errs = append(errs, fmt.Errorf("missing coverage for min_value(%d) in threshold value: %q", minValue, thresholds[0]))
```
Taken care of in the refactor.

shrutiparabgoogle commented:
Thanks you for the suggestion @seaneganx. I refactored with mostly the same structure as suggested, except for a few renames. Also, added a testcase for the case you caught in one of the comments.

seaneganx commented on cmd/registry/scoring/parser.go:
```diff
@@ -101,36 +101,72 @@ func validateScoreFormula(targetName patterns.ResourceName, scoreFormula *rpc.Sc
+
+		// Check for out of limits conditions
+		if rangeMax > maxValue || rangeMax < minValue {
+			errs = append(errs, fmt.Errorf("invalid range.max [%d, %d]: range.max(%d) should be within min_value(%d) and max_value(%d) limits", rangeMin, rangeMax, rangeMax, minValue, maxValue))
```
Should this say "invalid range" like it does below?

shrutiparabgoogle replied:
Done.

shrutiparabgoogle commented on cmd/registry/scoring/parser.go:
```diff
@@ -0,0 +1,160 @@
+	errs := make([]error, 0, len(patternErrs)+len(formulaErrs)+len(thresholdErrs))
+	errs = append(errs, patternErrs...)
+	errs = append(errs, formulaErrs...)
+	errs = append(errs, thresholdErrs...)
```
Done.

I started with the separate lists approach because there was a dependency on checking the formula based on the errors found/not found in `target_resource.pattern`. That dependency can still be taken care of by maintaining only one list of `totalErrs`.
